src/lib.rs

```diff
@@ -621,40 +647,111 @@ impl<'a> GzipLayerWriter<'a> {
-    fn write(&mut self, srcbuf: &[u8]) -> std::io::Result<usize> {
-        self.compressor.get_mut().clear();
-        self.compressor.write_all(srcbuf).unwrap();
-        self.uncompressed_hash.update(srcbuf)?;
-        let compressed_buf = self.compressor.get_mut().as_slice();
-        self.bw.write_all(compressed_buf)?;
-        Ok(srcbuf.len())
+    fn write(&mut self, data: &[u8]) -> std::io::Result<usize> {
+        self.0.write(data)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.0.flush()
+    }
+}
+
+#[cfg(feature = "zstd")]
+impl<'a> ZstdLayerWriter<'a> {
+    /// Create a writer for a gzip compressed layer blob.
+    fn new(ocidir: &'a Dir, c: Option<i32>) -> Result<Self> {
+        let bw = BlobWriter::new(ocidir)?;
+        let encoder = zstd::Encoder::new(bw, c.unwrap_or(0))?;
+        Ok(Self(Sha256Writer::new(encoder)))
+    }
+
+    /// Consume this writer, flushing buffered data and put the blob in place.
+    pub fn complete(self) -> Result<Layer> {
+        let (uncompressed_sha256, enc) = self.0.finish();
+        let blob = enc.finish()?.complete()?;
+        Ok(Layer {
+            blob,
+            uncompressed_sha256,
+            media_type: MediaType::ImageLayerZstd,
+        })
+    }
+}
+
+#[cfg(feature = "zstdmt")]
+impl<'a> ZstdLayerWriter<'a> {
+    /// Create a writer for a zstd compressed layer blob, with multithreaded compression enabled.
+    ///
+    /// The `n_workers` parameter specifies the number of threads to use for compression, per
+    /// [Encoder::multithread]]
+    fn multithread(ocidir: &'a Dir, c: Option<i32>, n_workers: u32) -> Result<Self> {
+        let bw = BlobWriter::new(ocidir)?;
+        let mut encoder = zstd::Encoder::new(bw, c.unwrap_or(0))?;
+        encoder.multithread(n_workers)?;
+        Ok(Self(Sha256Writer::new(encoder)))
+    }
+}
+
+#[cfg(feature = "zstd")]
+impl<'a> std::io::Write for ZstdLayerWriter<'a> {
+    fn write(&mut self, data: &[u8]) -> std::io::Result<usize> {
+        self.0.write(data)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.0.flush()
+    }
+}
+
+/// Wraps a writer and calculates the sha256 digest of data written to the inner writer
+struct Sha256Writer<W> {
+    inner: W,
+    sha: openssl::sha::Sha256,
+}
+
+impl<W> Sha256Writer<W> {
+    pub(crate) fn new(inner: W) -> Self {
+        Self {
+            inner,
+            sha: openssl::sha::Sha256::new(),
+        }
+    }
+
+    /// Return the hex encoded sha256 digest of the written data, and the underlying writer
+    pub(crate) fn finish(self) -> (Sha256Digest, W) {
+        let digest = hex::encode(self.sha.finish());
+        (Sha256Digest::from_str(&digest).unwrap(), self.inner)
+    }
+}
+
+impl<W> Write for Sha256Writer<W>
+where
+    W: Write,
+{
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        let len = self.inner.write(buf)?;
+        self.sha.update(&buf[..len]);
```

Comment:
Ah you know what, this is clearly better code than what it replaces. We need to handle short writes, and I think that's what I messed up [in the original](https://github.com/containers/ocidir-rs/pull/24/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759L648).